README
------------------------------------------------------------------------------

  Documentation of lv is provided on the web. See the following URL:

    http://www.ff.iij4u.or.jp/~nrt/lv/

------------------------------------------------------------------------------

   All rights reserved. Copyright (C) 1996-2005 by NARITA Tomio.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

   See also GNU General Public License Version 2, included in this
   archive as GPL.txt.

------------------------------------------------------------------------------

INSTALLATION:

    0) Expand lv archive, using gunzip/tar.
    1) Change your working directory to ``(extracted sub directory)/build''.
    2) Execute ``../src/configure'' to configure compiler flags.
    3) Launch ``make''.
    4) Then, launch ``make install'' as root.

------------------------------------------------------------------------------
NARITA Tomio
email: nrt@ff.iij4u.or.jp
Homepage: http://www.ff.iij4u.or.jp/~nrt/


debian/changelog
lv (4.51-2.3) unstable; urgency=medium

  * Non-maintainer upload.
  * replace sigvec usage (Closes: #769546).
    - debian/patches/sigvec.patch: new file. thanks to Florian Weimer
      <fweimer@redhat.com>.
    - debian/rules: enable autoreconf add-on.
    - debian/autoreconf: set subdir.
    - debian/control: Depends: dh-autoreconf.
  * bump up Standards-Version 3.9.6

 -- HIGUCHI Daisuke (VDR dai) <dai@debian.org>  Mon, 27 Apr 2015 21:43:39 +0900

lv (4.51-2.2) unstable; urgency=medium

  * Non-maintainer upload.
  * debian/control, debian/rules, debian/compat: use dh9.
  * debian/control
    - add Vcs-* tags.
    - add Homepage: tag.
    - add ${misc:Depends} to Depends:.
    - add xz-utils to Recommends:.
  * debian/source/format: set 3.0 (quilt).
  * debian/patches/*: rename from debian/patch* and add DEP-3 headers.
  * debian/copyright: convert to DEP-5.
  * debian/patches/fix-hyphen-used-as-minus-sign.diff: new file.
  * debian/lv.doc-base: new file.
  * bump up Standards-Version 3.9.5

 -- HIGUCHI Daisuke (VDR dai) <dai@debian.org>  Sat, 15 Mar 2014 03:42:28 +0900

lv (4.51-2.1) unstable; urgency=low

  * Non-maintainer upload.
  * Add 1 patch:
    - debian/@patch.lv.660358.diff: support automatic xz and lzma
      decompression. thanks to "A. N. Other" <a.n.other.debian@gmail.com>.
      (Closes: #660358)
  * debian/watch: Add. thanks to Bart Martens <bartm@debian.org>.

 -- HIGUCHI Daisuke (VDR dai) <dai@debian.org>  Thu, 03 Oct 2013 21:42:15 +0900

lv (4.51-2) unstable; urgency=low

  * Add 3 (+1) patches:
    - debian/@patch.lv.20050502_~+num-pat-option: Add +num and +/pattern
      option.
    - debian/@patch.lv.20050506_%src%file.c_~enable-fastio-use-fread: Improve
      large file processing performance over typically +40% to +300%.
    - debian/patch.lv.misc: Enable above changes.
    - debian/@patch.lv.20051030_%src%_~use_off_t: Update copyright year to
      2005.  Support large file.  (Closes: #173060)
    - debian/rules: Likewise.

 -- GOTO Masanori <gotom@debian.org>  Fri,  6 May 2005 15:05:17 +0900

lv (4.51-1) unstable; urgency=low

  * New upstream release.  (Closes: #291528)
  * Change maintainer address.

 -- GOTO Masanori <gotom@debian.org>  Mon,  2 May 2005 17:43:31 +0900

lv (4.50-1) unstable; urgency=low

  * New upstream release.  (Closes: #220716)
  * Lgrep segv problem is fixed when lgrep greps through many files
    (Closes: #188288)
  * Wrong ISO-2022 final byte is fixed.  (Closes #207718)

 -- GOTO Masanori <gotom@debian.or.jp>  Sun, 16 Nov 2003 01:21:59 +0900

lv (4.49.5-2) unstable; urgency=high

  * rebuild 4.49.5-1 under libc6 2.3.1-17.  4.49.5-1 was built under
    unexisted libc6 version.

 -- GOTO Masanori <gotom@debian.or.jp>  Wed, 14 May 2003 22:15:46 +0900

lv (4.49.5-1) unstable; urgency=high

  * lv has possible local root exploit problem.  lv always reads .lv file
    in the current directory.  So if ill-intentioned user could put an .lv
    with the line '-E/path/to/malicious/program', any user who uses the v
    (edit) command in that directory will unwittingly execute the malicious
    program; this could be exploited to run an arbitrary script with root
    permissions.  This version fixes the problem.  Bug found by 
    Leonard Stiles <ljs@uk2.net>.  (#190941).
  * lv does not check environment variable EDITOR or VISUAL, it does not 
    follow debian policy 12.4.  This version fixes this problem.
    Patched by Michael Weber <michaelw@debian.org>.  (Closes: #181402)
  * Fix 'unaligned access to' message from kernel due to unaligned 
    access on 64 bit architecture machine.

 -- GOTO Masanori <gotom@debian.or.jp>  Mon, 12 May 2003 17:20:38 +0900

lv (4.49.4-8) unstable; urgency=low

  * Fix 'problem when stdin isn't tty and there are file options'.
    Patched by Fumitoshi UKAI <ukai@debian.or.jp>. Closes: #127861

 -- GOTO Masanori <gotom@debian.or.jp>  Mon,  7 Oct 2002 10:40:23 +0900

lv (4.49.4-7) unstable; urgency=low

  * debian/control: fix typo in description field.

 -- GOTO Masanori <gotom@debian.or.jp>  Wed, 21 Nov 2001 09:24:26 +0900

lv (4.49.4-6) unstable; urgency=low

  * lv.1: fix .SH NAME format. when you used apropos and man -k, the
    error was displayed (closes: #119125).

 -- GOTO Masanori <gotom@debian.or.jp>  Mon, 12 Nov 2001 00:55:42 +0900

lv (4.49.4-5) unstable; urgency=low

  * append debian/postinst, debian/prerm script (I forgot!). these scripts 
    register pager alternatives.

 -- GOTO Masanori <gotom@debian.or.jp>  Wed,  7 Nov 2001 16:41:09 +0900

lv (4.49.4-4) unstable; urgency=low

  * debian/control: add build-dependency libncurses5-dev (closes: #118437).

 -- GOTO Masanori <gotom@debian.or.jp>  Tue,  6 Nov 2001 23:23:50 +0900

lv (4.49.4-3) unstable; urgency=low

  * applied the patch which forces displaying text whether @TERM is 
    valid or not (X-ML-Name: debian-users @debian.or.jp, X-Mail-Count: 30379).

 -- GOTO Masanori <gotom@debian.or.jp>  Tue,  6 Nov 2001 01:09:03 +0900

lv (4.49.4-2) unstable; urgency=low

  * Update Standards-Version: 3.5.2.
  * I believe Bugs#102893 is resolved in this release
    (I guess that SPARC64 build environment have some unmatch of lv
    and its dependency library) (Closes: #102893).
  * Experimental patch version 4.49.4.1.
    This patch supports ISO-8859-10,13,14,15, Copyright update.

 -- GOTO Masanori <gotom@debian.or.jp>  Mon,  5 Nov 2001 23:08:19 +0900

lv (4.49.4-1) unstable; urgency=low

  * New Upstream Version.

 -- GOTO Masanori <gotom@debian.or.jp>  Wed, 27 Sep 2000 03:03:46 +0900

lv (4.49.3-5) frozen unstable; urgency=low

  * src/command.c: Deleted comments of line 187 and added key bind of
    'DEL'. Reported by John Belmonte <john@nanaon-sha.co.jp>,
    patched by Tomohiro KUBOTA <kubota@debian.or.jp>. (closes: #66054)

 -- GOTO Masanori <gotom@debian.or.jp>  Thu, 22 Jun 2000 23:06:34 +0900

lv (4.49.3-4) frozen unstable; urgency=low

  * Changelog from lv 4.49.3-2 and 4.49.3-3, both of which were never
    officially released.

src/utf.c
/*
 * utf.c
 *
 * All rights reserved. Copyright (C) 1996 by NARITA Tomio.
 * $Id: utf.c,v 1.7 2004/01/05 07:23:29 nrt Exp $
 */
/*
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <stdio.h>

#include <import.h>
#include <decode.h>
#include <escape.h>
#include <encode.h>
#include <unirev.h>
#include <begin.h>
#include <utf.h>

private int base64char[ 64 ] = {
  'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H',
  'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P',
  'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X',
  'Y', 'Z', 'a', 'b', 'c', 'd', 'e', 'f',
  'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n',
  'o', 'p', 'q', 'r', 's', 't', 'u', 'v',
  'w', 'x', 'y', 'z', '0', '1', '2', '3',
  '4', '5', '6', '7', '8', '9', '+', '/'
};

private int base64[ 256 ] = {
/* 0   1   2   3   4   5   6   7   8   9   a   b   c   d   e   f */
  -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* 0 */
  -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* 1 */
  -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 62, -1, -1, -1, 63, /* 2 */
  52, 53, 54, 55, 56, 57, 58, 59, 60, 61, -1, -1, -1, -1, -1, -1, /* 3 */
  -1,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, /* 4 */
  15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, -1, -1, -1, -1, -1, /* 5 */
  -1, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, /* 6 */
  41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, -1, -1, -1, -1, -1, /* 7 */
  -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* 8 */
  -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* 9 */
  -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* a */
  -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* b */
  -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* c */
  -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* d */
  -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* e */
  -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1  /* f */
};

private void DecodeAddUnicode( byte attr, ic_t uni )
{
  if( uni <= (ic_t)DEL ){
    if( uni <= (ic_t)SP ){
      if( (ic_t)SP == uni )
	DecodeAddSpace( attr );
      else
	DecodeAddControl( (byte)uni );
    } else if( uni < (ic_t)DEL )
      DecodeAddIchar( ASCII, uni, attr );
    else {
      DecodeAddControl( (byte)uni );
    }
  } else
    DecodeAddIchar( UNICODE, uni, attr );
}

private char DecodeBase64( byte attr, byte ch )
{
  unsigned long acc;
  int val;
  ic_t uni;

  do {
    if( 0 <= (val = base64[ (int)ch ]) ){
      acc = val << 26;
      GetChar( ch );
      if( 0 <= (val = base64[ (int)ch ]) ){
	acc |= val << 20;
	GetChar( ch );
	if( 0 <= (val = base64[ (int)ch ]) ){
	  acc |= val << 14;
	  uni = acc >> 16;
	  DecodeAddUnicode( attr, uni );
	  acc = ( acc & 0x0000ffffL ) << 16;
	  GetChar( ch );
	  if( 0 <= (val = base64[ (int)ch ]) ){
	    acc |= val << 24;
	    GetChar( ch );
	    if( 0 <= (val = base64[ (int)ch ]) ){
	      acc |= val << 18;
	      GetChar( ch );
	      if( 0 <= (val = base64[ (int)ch ]) ){
		acc |= val << 12;
		uni = acc >> 16;
		DecodeAddUnicode( attr, uni );
		acc = ( acc & 0x0000ffffL ) << 16;
		GetChar( ch );
		if( 0 <= (val = base64[ (int)ch ]) ){
		  acc |= val << 22;
		  GetChar( ch );
		  if( 0 <= (val = base64[ (int)ch ]) ){
		    acc |= val << 16;
		    uni = acc >> 16;
		    DecodeAddUnicode( attr, uni );
		    GetChar( ch );
		  }
		}
	      }
	    }
	  }
	}
      }
    }
  } while( val >= 0 );

  return ch;
}

public void DecodeUTF( state_t *state, byte codingSystem )
{
  byte ch;
  ic_t uni;

  for( ; ; ){
    GetChar( ch );
utfContinue:
    if( ch <= SP ){
      if( SP == ch )
	DecodeAddSpace( state->attr );
      else if( ESC == ch ){
	if( FALSE == DecodeEscape( state ) )
	  break;
      } else if( HT == ch )
	DecodeAddTab( state->attr );
      else if( BS == ch )
	DecodeAddBs();
      else
	DecodeAddControl( ch );
    } else {
      if( ch < (ic_t)DEL ){
	if( UTF_8 == codingSystem ){
	  DecodeAddIchar( ASCII, (ic_t)ch, state->attr );
	} else {
	  /* UTF_7 */
	  if( '+' != ch ){
	    DecodeAddIchar( ASCII, (ic_t)ch, state->attr );
	  } else {
	    /* '+' == ch */
	    GetChar( ch );
	    if( '-' == ch ){
	      DecodeAddIchar( ASCII, (ic_t)'+', state->attr );
	    } else {
	      ch = DecodeBase64( state->attr, ch );
	      if( SHIGH == SIDX || LF == ch || CR == ch )
		return;
	      else if( '-' == ch )
		continue;
	      else {
		decoding_penalty++;
		goto utfContinue;
	      }
	    }
	  }
	}
      } else if( 0xc0 == ( 0xe0 & ch ) ){
	/* 1st of 2 */
	uni = (ic_t)( 0x1f & ch ) << 6;
	GetChar( ch );
	if( 0x80 == ( 0xc0 & ch ) ){
	  /* trailor */
	  uni |= (ic_t)( 0x3f & ch );
	  DecodeAddUnicode( state->attr, uni );
	}
      } else if( 0xe0 == ( 0xf0 & ch ) ){
	/* 1st of 3 */
	uni = (ic_t)( 0x0f & ch ) << 12;
	GetChar( ch );
	if( 0x80 == ( 0xc0 & ch ) ){
	  /* trailor */
	  uni |= (ic_t)( 0x3f & ch ) << 6;
	  GetChar( ch );
	  if( 0x80 == ( 0xc0 & ch ) ){
	    /* trailor */
	    uni |= (ic_t)( 0x3f & ch );
	    if( uni != UNICODE_BOM )
	      DecodeAddUnicode( state->attr, uni );
	  }
	}
      } else {
	decoding_penalty++;
	DecodeAddControl( ch );
      }
    }
  }
}

typedef enum { ASCII_STATE, PENDING_NONE, PENDING_4, PENDING_2 } utf7_state_t;

private void EncodeUTF7PendingBit( utf7_state_t stat, int attr, ic_t uni )
{
  if( PENDING_NONE != stat )
    EncodeAddCharAbsolutely( attr, base64char[ uni ] );
  EncodeAddCharAbsolutely( attr, (ic_t)'-' );
}

public void EncodeUTF7( i_str_t *istr, int head, int tail,
		       byte codingSystem, boolean_t binary )
{
  int idx, attr, lastAttr;
  ic_t ic, uni = 0;
  byte cset;
  utf7_state_t stat;

  attr = lastAttr = 0;
  stat = ASCII_STATE;

  for( idx = head ; idx < tail ; idx++ ){
    cset = istr[ idx ].charset;
    ic = istr[ idx ].c;
    attr = (int)istr[ idx ].attr << 8;
    if( attr != lastAttr ){
      if( ASCII_STATE != stat ){
	EncodeUTF7PendingBit( stat, lastAttr, uni );
	stat = ASCII_STATE;
      }
      lastAttr = attr;
    }
    if( cset < PSEUDO ){
      if( ASCII == cset ){
	if( ASCII_STATE != stat ){
	  EncodeUTF7PendingBit( stat, lastAttr, uni );
	  stat = ASCII_STATE;
	}
	if( (ic_t)'+' == ic ){
	  EncodeAddCharAbsolutely( attr, ic );
	  EncodeAddCharAbsolutely( attr, (ic_t)'-' );
	} else {
	  EncodeAddChar( attr, ic );
	}
      } else {
	if( UNICODE != cset )
	  ic = RevUNI( ic, &cset );
	if( ic < 0x80U ){
	  if( ASCII_STATE != stat ){
	    EncodeUTF7PendingBit( stat, lastAttr, uni );
	    stat = ASCII_STATE;
	  }
	  if( (ic_t)'+' == ic ){
	    EncodeAddCharAbsolutely( attr, ic );
	    EncodeAddCharAbsolutely( attr, (ic_t)'-' );
	  } else {
	    EncodeAddChar( attr, ic );
	  }
	} else {
	  switch( stat ){
	  case ASCII_STATE:
	    EncodeAddChar( attr, (ic_t)'+' );
	  case PENDING_NONE:
	    EncodeAddChar( attr, base64char[ ic >> 10 ] );
	    EncodeAddChar( attr, base64char[ 0x3f & ( ic >> 4 ) ] );
	    uni = ( ic & 0x0f ) << 2;
	    stat = PENDING_4;
	    break;
	  case PENDING_4:
	    uni |= ic >>14;
	    EncodeAddChar( attr, base64char[ uni ] );
	    EncodeAddChar( attr, base64char[ 0x3f & ( ic >> 8 ) ] );
	    EncodeAddChar( attr, base64char[ 0x3f & ( ic >> 2 ) ] );
	    uni = ( ic & 0x03 ) << 4;
	    stat = PENDING_2;
	    break;
	  case PENDING_2:
	    uni |= ic >> 12;
	    EncodeAddChar( attr, base64char[ uni ] );
	    EncodeAddChar( attr, base64char[ 0x3f & ( ic >> 6 ) ] );
	    EncodeAddChar( attr, base64char[ 0x3f & ic ] );
	    stat = PENDING_NONE;
	  }
	}
      }
    } else {
      if( ASCII_STATE != stat ){
	EncodeUTF7PendingBit( stat, lastAttr, uni );
	stat = ASCII_STATE;
      }
      if( FALSE == EncodeAddPseudo( attr, ic, cset, binary ) ){
	break;
      }
    }
  }
  if( ASCII_STATE != stat ){
    EncodeUTF7PendingBit( stat, lastAttr, uni );
  }
}

public void EncodeUTF8( i_str_t *istr, int head, int tail,
		       byte codingSystem, boolean_t binary )
{
  int idx, attr;
  ic_t ic;
  byte cset;

  for( idx = head ; idx < tail ; idx++ ){
    cset = istr[ idx ].charset;
    ic = istr[ idx ].c;
    attr = (int)istr[ idx ].attr << 8;
    if( cset < PSEUDO ){
      if( ASCII == cset ){
	EncodeAddChar( attr, ic );
      } else {
	if( UNICODE != cset )
	  ic = RevUNI( ic, &cset );
	if( ic < 0x80U ){
	  EncodeAddChar( attr, ic );
	} else if( ic < 0x0800U ){
	  EncodeAddChar( attr, 0xc0 | ( ic >> 6 ) );
	  EncodeAddChar( attr, 0x80 | ( 0x3f & ic ) );
	} else {
	  EncodeAddChar( attr, 0xe0 | ( ic >> 12 ) );
	  EncodeAddChar( attr, 0x80 | ( 0x3f & ( ic >> 6 ) ) );
	  EncodeAddChar( attr, 0x80 | ( 0x3f & ( ic ) ) );
	}
      }
    } else if( FALSE == EncodeAddPseudo( attr, ic, cset, binary ) ){
      break;
    }
  }
}

#ifdef USE_UTF16
public void DecodeUTF16( state_t *state, byte codingSystem )
{
  byte ch1, ch2; /* hibyte, lobyte */
  ic_t uni;

  for( ; ; ){
    if( UTF_16 == codingSystem ){
      GetChar( ch1 );
      GetChar( ch2 );
      uni = (ic_t)(( ch1 << 8 ) | ch2);
      if( uni == UNICODE_BOM ){
	codingSystem = UTF_16BE;
	continue;
      }
      uni = (ic_t)(( ch2 << 8 ) | ch1);
      if( uni == UNICODE_BOM ){
	codingSystem = UTF_16LE;
	continue;
      }
      /* fallthru assume UTF_16BE */
    } else if( UTF_16LE == codingSystem ){
      GetChar( ch2 );
      GetChar( ch1 );
    } else {
      GetChar( ch1 );
      GetChar( ch2 );
    }

    if( ch1 == 0 && ch2 <= SP ){
      if( SP == ch2 )
	DecodeAddSpace( state->attr );
#if 0
      else if( ESC == ch2 ){
	if( FALSE == DecodeEscape( state ) )
	  break;
      }
#endif
      else if( HT == ch2 )
	DecodeAddTab( state->attr );
      else if( BS == ch2 )
	DecodeAddBs();
      else
	DecodeAddControl( ch2 );
    } else if( ch1 == 0 && ch2 < (ic_t)DEL ){
      DecodeAddIchar( ASCII, (ic_t)ch2, state->attr );
    } else {
      uni = (ic_t)(( ch1 << 8 ) | ch2);
      if( uni != UNICODE_BOM )
	DecodeAddUnicode( state->attr, uni );
    }
  }
}

public void EncodeUTF16( i_str_t *istr, int head, int tail,
			byte codingSystem, boolean_t binary )
{
  int idx, attr;
  ic_t ic;
  byte cset;

  for( idx = head ; idx < tail ; idx++ ){
    cset = istr[ idx ].charset;
    ic = istr[ idx ].c;
    attr = (int)istr[ idx ].attr << 8;
    if( cset < PSEUDO ){
      if( ASCII == cset ){
	if( UTF_16LE == codingSystem ){
	  EncodeAddChar( attr, ic );
	  EncodeAddChar( attr, 0 );
	} else {
	  EncodeAddChar( attr, 0 );
	  EncodeAddChar( attr, ic );
	}
      } else {
	if( UNICODE != cset )
	  ic = RevUNI( ic, &cset );
	if( UTF_16LE == codingSystem ){
	  EncodeAddChar( attr, 0xff & ic );
	  EncodeAddChar( attr, ic >> 8 );
	} else {
	  EncodeAddChar( attr, ic >> 8 );
	  EncodeAddChar( attr, 0xff & ic );
	}
      }
    } else if( FALSE == EncodeAddPseudo16( attr, ic, cset, binary,
					  UTF_16LE == codingSystem ) ){
      break;
    }
  }
}
#endif /* USE_UTF16 */
